app/attributes/fieldvalidator.cpp

```diff
@@ -0,0 +1,259 @@
+      level = Error;
+      return false;
+    }
+  }
+  return true;
+}
+
+bool FieldValidator::validateNumericField( const FormItem &item, QVariant &value, QString &validationMessage, ValidationMessageLevel &level )
+{
+  const QgsField field = item.field();
+
+  if ( value.isNull() )
+  {
+    return true;
+  }
+
+  QString errorMessage;
+
+  bool containsDecimals = value.toString().contains( QLocale().decimalPoint() ) || value.toString().contains( "." );
+
+  if ( !field.convertCompatible( value, &errorMessage ) )
+  {
+    if ( errorMessage.contains( QStringLiteral( "too large" ) ) )
+    {
+      validationMessage = Resources::Texts::Validation::numberExceedingVariableLimits;
+    }
+    else
+    {
+      validationMessage = Resources::Texts::Validation::numberInvalid;
+    }
+
+    level = Error;
+    return false;
+  }
+  else if ( containsDecimals && field.type() != QVariant::Type::Double )
+  {
+    /* ConverCompatible check passes for doubles written into int fields,
+     * however, the value would not be saved and would get replaced by zero,
+     * so we need to handle it here and set invalid state for such input.
+     */
+    validationMessage = Resources::Texts::Validation::numberMustBeInt;
+    level = Error;
+    return false;
+  }
+
+  bool isRangeEditable = item.editorWidgetType() == QStringLiteral( "Range" ) &&
+                         item.editorWidgetConfig().value( QStringLiteral( "Style" ) ) == QStringLiteral( "SpinBox" );
+
+  // Check min/max range
+  if ( isRangeEditable )
+  {
+    double min = item.editorWidgetConfig().value( "Min" ).toDouble();
+    double max = item.editorWidgetConfig().value( "Max" ).toDouble();
+    double val = value.toDouble();
+
+    if ( val < min )
+    {
+      validationMessage = Resources::Texts::Validation::numberLowerBoundReached.arg( min );
+      level = Error;
+      return false;
+    }
+    else if ( val > max )
+    {
+      validationMessage = Resources::Texts::Validation::numberUpperBoundReached.arg( max );
+      level = Error;
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool FieldValidator::validateGenericField( const FormItem &item, QVariant &value, QString &validationMessage, ValidationMessageLevel &level )
+{
+  const QgsField field = item.field();
+
+  if ( !field.convertCompatible( value ) )
+  {
+    validationMessage = Resources::Texts::Validation::genericValidationFailed;
+    level = Error;
+    return false;
+  }
+
+  return true;
+}
+
+QString FieldValidator::constructConstraintValidationMessage( const FormItem &item, const QStringList &unmetConstraints )
+{
+  const QgsField field = item.field();
+  QgsFieldConstraints fldCons = field.constraints();
+  QStringList validationMessages;
+
+  bool hasNotNullConstraint = fldCons.constraints() & QgsFieldConstraints::ConstraintNotNull;
+  bool notNullViolated = unmetConstraints.contains( QStringLiteral( "value is NULL" ) );
```

Comment:
these messages from qgis are translatable unfortunately, but as long as we don't ship translations for qgis code, it should work :-) maybe at some point we should just make an internal copy of QgsVectorLayerUtils::validateAttribute() to provide "our" texts (or fix the function in qgis to provide better messages)